qcodes/instrument_drivers/Lakeshore/lakeshore_base.py

```diff
@@ -0,0 +1,549 @@
+from bisect import bisect
+
+import numpy as np
+
+from qcodes import VisaInstrument, InstrumentChannel, ChannelList
+from qcodes.instrument.group_parameter import GroupParameter, Group
+from qcodes.utils import validators as vals
+
+
+log = logging.getLogger(__name__)
+
+
+class BaseOutput(InstrumentChannel):
+    """
+    Base class for the outputs of Lakeshore temperature controllers
+
+    Args:
+        parent
+            instrument that this channel belongs to
+        output_name
+            name of this output
+        output_index
+            identifier for this output that is used in VISA commands of the
+            instrument
+        has_pid
+            if True, then the output supports closed loop control,
+            hence it will have three parameters to set it up: 'P', 'I', and 'D'
+    """
+
+    MODES: ClassVar[Dict[str, int]] = {}
+    RANGES: ClassVar[Dict[str, int]] = {}
+
+    def __init__(self, parent, output_name, output_index, had_pid: bool=True) \
+            -> None:
+        super().__init__(parent, output_name)
+
+        self.INVERSE_RANGES: Dict[int, str] = {
+            v: k for k, v in self.RANGES.items()}
+
+        self._has_pid = had_pid
+        self._output_index = output_index
+
+        self.add_parameter('mode',
+                           label='Control mode',
+                           docstring='Specifies the control mode',
+                           val_mapping=self.MODES,
+                           parameter_class=GroupParameter)
+        self.add_parameter('input_channel',
+                           label='Input channel',
+                           docstring='Specifies which measurement input to '
+                                     'control from (note that only '
+                                     'measurement inputs are available)',
+                           get_parser=int,
+                           parameter_class=GroupParameter)
+        self.add_parameter('powerup_enable',
+                           label='Power-up enable on/off',
+                           docstring='Specifies whether the output remains on '
+                                     'or shuts off after power cycle.',
+                           val_mapping={True: 1, False: 0},
+                           parameter_class=GroupParameter)
+        self.output_group = Group([self.mode, self.input_channel,
+                                   self.powerup_enable],
+                                  set_cmd=f'OUTMODE {output_index}, {{mode}}, '
+                                          f'{{input_channel}}, '
+                                          f'{{powerup_enable}}',
+                                  get_cmd=f'OUTMODE? {output_index}')
+
+        # Parameters for Closed Loop PID Parameter Command
+        if self._has_pid:
+            self.add_parameter('P',
+                               label='P value of closed-loop controller',
+                               docstring='The value for control loop '
```

Comment:
In the docstring, refer to the manual at the right page.

Reply:
as discussed, i'm going to do it whenever necessary. (added to todo list so that i dont forget)